features/0793-unqualfied-dids-transition/README.md

```diff
@@ -0,0 +1,102 @@
+# Aries RFC 0793: Unqualified DID Transition
+
+- Authors: [Sam Curren](mailto:swcurran@cloudcompass.ca)
+- Status: [PROPOSED](/README.md#proposed)
+- Since: 2023-07-11
+- Status Note: In Pre-work **Target Completion Date: 2023.10.15** 
+- Supersedes:
+- Start Date: 2023-07-11
+- Tags: [feature](/tags.md#feature), [community-update](/tags.md#community-update)
+
+## Summary
+
+Historically, Aries use of the Indy SDK's wallet included the use of 'unqualified DIDs' or DIDs without a did: prefix and method. 
+This transition documents the process of migrating any such DIDs still in use to fully qualified DIDs.
+
+The transition from unqualified DIDs to peer:did:2 is described here: https://github.com/TimoGlastra/legacy-did-transformation
+
+The use and support for did:peer:3 is also included in this transition.
+
+This process ONLY applies to agent codebases using unqualified DIDs. Any codebases not using unqualified DIDs may ignore this process.
+
+This RFC follows the guidance in [RFC
+0345](../../concepts/0345-community-coordinated-update/README.md) about
+community-coordinated updates to (try to) ensure that independently deployed,
+interoperable agents remain interoperable throughout this transition.
+
+The transition from the unqualified to qualified DIDs will occur in four steps:
+
+- **Pre-work**: where we agree on the transition plan outlined in this RFC.
+  - Finalize did:peer:3 details, including feature discovery support of did:peer:3.
```

Comment:
The 'feature discovery' part seems quite important to me but couldn't find further explanations about how it should be achieved, since DIDs are used from the start on connection protocols. And we have to consider that not only did:peer:3 support must be discovered but also did:peer:2 (and did:peer in general I would say), as AFAIK not every framework out there supports it at the moment.

Can be discovered by using an extra field on Out of Band invitation messages (or an existing one like `accept`)?

Could it be done using a recursive strategy where the Agent first tries to send a DID Exchange Request using a did:peer:2 and:
- If connection is successfully established, use Discover Features protocol to ask for did:peer:3 support (a `feature-type` must be created for that purpose) and both parties will use did:peer:3 if so 
- If the request is rejected (Problem Report sent with `request_not_accepted`) try again using a did:peer:1 (and eventually unqualified DID if the error persists) 

Could be done by sending did:peer:3 right from the start (passing also the entire DIDDoc) and if the request is accepted simply use did:peer:3 from then (and if not follow the recursive strategy falling back to other DID methods?)

Reply:
Just a few more questions after today's Aries WG call: using Discover Features V2 `feature-type` to discover `did:peer:3` variant support seems quite straightforward in existing connections but It's still not clear to me how to handle the Aries community transition cases for establishing new DIDComm relationships in some situations:

- OOB invitation messages that contain a public DID, or direct interactions to a public DID (in DIDComm V2). How do we know we can use a `did:peer:2` to interact with it?
- OOB invitation messages including a `from` (2.0) or `services` (1.1) field that has a `did:peer:2`. This is good for the invitee to know beforehand the inviter supports did:peer:2 so this problem is solved. But... what if the invitee does not support `did:peer:2`? Do we need to create different invitation messages for legacy agents and the new ones?

I guess at a first step we can start all interactions using unqualified DIDs and do the feature discovery, so we'd use the legacy invitation message for everyone until we are sure everybody transitioned to `did:peer` support. From then on we start using the new fancy invitations with did:peer:2 (or 3?) from the start.

However, I would not assume that every agent out there is still using unqualified DIDs for establishing connections. In AFJ for instance we are using did:peer:1 in DID Exchange.

Reply:
The idea is for the transition is that ASAP, all the libraries be able to handle receiving either unqualified p2p DIDs (old style we have been using for years) or did:peer:2 DIDs for establishing a connection, and unqualified or did:peer:3 DIDs for interactions using the connection.  In doing this, a did:peer:3 should be defined for every unqualified DID.  During this period, all deployments are expected to initiate new connections with the current unqualified p2p DIDs.

After a set date TBD, all deployments get updated to send out only did:peer:2 for establishing connections, and only send did:peer:3 for interactions using the connections. Obivously, the “ASAP” in the previous paragraph **MUST** be completed by all deployments by this magic date.

On your later point about — what if the invitee does not support `did:peer:2`? When creating a new connection, there is no way to know, as you can’t run `discover-features` without a connection.  That is why the community coordinated update is done.  “All” deployments **MUST** support handling `did:peer:2` by the date and so all deployments assume they will.  If the deployments haven’t been updated — that’s unfortunate.

So legacy invitation before the date, anytime after, switch the “new fancy invitations” with did:peer:2 (can’t be did:peer:3 because the invitation must include the needed DIDDoc information).  Once the connection is established, everyone can use the equivalent did:peer:3.

Does that clarify it?